core/trino-main/src/main/java/io/trino/sql/planner/ScopeAware.java

```diff
@@ -109,25 +129,18 @@ private Boolean scopeAwareComparison(Node left, Node right)
                 ResolvedField leftField = analysis.getResolvedField(leftExpression);
                 ResolvedField rightField = analysis.getResolvedField(rightExpression);
 
-                Scope leftScope = leftField.getScope();
-                Scope rightScope = rightField.getScope();
+                boolean leftFieldInSubqueryScope = leftField.getScope().hasOuterParent(queryScope);
+                boolean rightFieldInSubqueryScope = rightField.getScope().hasOuterParent(queryScope);
 
                 // For subqueries of the query associated with the current expression, compare by syntax
-                // Note: it'd appear that hash() and equal() are inconsistent with each other in the case that:
-                //    * left.hasOuterParent(...) == true and right.hasOuterParent(...) == false
-                //    * leftField.getFieldId().equals(rightField.getFieldId()) == true
-                // Both fields would seem to have different hashes but be equal to each other.
-                // However, this cannot happen because we *require* that both expressions being compared by
-                // rooted in the same "query scope" (i.e., sub-scopes that are local to each other) -- see ScopeAwareKey.equals().
-                // If both fields have the same field id, by definition they will produce the same result for hasOuterParent().
-                checkState(leftScope.hasOuterParent(queryScope) == rightScope.hasOuterParent(queryScope));
-                if (leftScope.hasOuterParent(queryScope) && rightScope.hasOuterParent(queryScope)) {
+                if (leftFieldInSubqueryScope && rightFieldInSubqueryScope) {
                     return treeEqual(leftExpression, rightExpression, CanonicalizationAware::canonicalizationAwareComparison);
                 }
-
-                // Otherwise, for references that come from the current query scope or an outer scope of the current
+                // For references that come from the current query scope or an outer scope of the current
                 // expression, compare by resolved field
-                return leftField.getFieldId().equals(rightField.getFieldId());
+                else {
+                    return leftField.getFieldId().equals(rightField.getFieldId());
+                }
```

Comment:
If we want to simplify the code this way, then we have to add a good explanatory comment to make it clear why it works.

I suggest something like:
```
// If any of the references comes from the current query scope or an outer scope
// of the current expression, compare by resolved field.
// * If both references come from the same scope, current or outer, we might
// identify that they are equal, even if they are syntactically different.
// * If the references come from different scopes, e.g. one from the current scope,
// and one from a subquery scope, the result is false.
```

Also, `else` is redundant.

Reply:
Thanks. One question about the following explanation:

> If both references come from the same scope, current or outer, we might identify that they are equal, even if they are syntactically different.

If both references come from outer, it's caught by the previous branch and compared syntactically?

Reply:
Oops. Sorry, I understood just before you submitting the comment. 🙇

I think your original code is easier to understand than the simplified version even if it has a little redundancy as you mentioned. So, I will revert my last change.